core/trino-main/src/main/java/io/trino/operator/exchange/LocalExchange.java

```diff
@@ -165,6 +192,50 @@ LocalExchangeSource getSource(int partitionIndex)
         return sources.get(partitionIndex);
     }
 
+    private static PartitionFunction createPartitionFunction(
+            PartitioningProviderManager partitioningProviderManager,
+            Session session,
+            BlockTypeOperators blockTypeOperators,
+            PartitioningHandle partitioning,
+            int partitionCount,
+            List<Type> partitionChannelTypes,
+            boolean isHashPrecomputed)
+    {
+        if (partitioning.getConnectorHandle() instanceof SystemPartitioningHandle) {
+            HashGenerator hashGenerator;
+            if (isHashPrecomputed) {
+                hashGenerator = new PrecomputedHashGenerator(0);
+            }
+            else {
+                hashGenerator = new InterpretedHashGenerator(partitionChannelTypes, IntStream.range(0, partitionChannelTypes.size()).toArray(), blockTypeOperators);
+            }
+            return new LocalPartitionGenerator(hashGenerator, partitionCount);
+        }
+
+        ConnectorNodePartitioningProvider partitioningProvider = partitioningProviderManager.getPartitioningProvider(partitioning.getConnectorId().get());
+        ConnectorBucketNodeMap connectorBucketNodeMap = partitioningProvider.getBucketNodeMap(
+                partitioning.getTransactionHandle().orElse(null),
+                session.toConnectorSession(),
+                partitioning.getConnectorHandle());
+        checkArgument(connectorBucketNodeMap != null, "No partition map %s", partitioning);
```

Comment:
When we register `ConnectorNodePartitioningProvider` in `PartitioningProviderManager` we should probaly wrap it so validation like this is performed internally as part of call to `getBucketNodeMap`

Reply:
It's based on preexisting code.